pkg/services/ngalert/provisioning/mute_timings.go

```diff
@@ -136,10 +147,20 @@ func (svc *MuteTimingService) UpdateMuteTiming(ctx context.Context, mt definitio
 		return definitions.MuteTimeInterval{}, nil
 	}
 
-	_, idx, err := getMuteTiming(revision, mt.Name)
+	old, idx, err := getMuteTiming(revision, mt.Name)
 	if err != nil {
 		return definitions.MuteTimeInterval{}, err
 	}
+
+	if mt.Version != "" {
+		curVersion := calculateMuteTimeIntervalFingerprint(old)
+		if curVersion != mt.Version {
+			return definitions.MuteTimeInterval{}, ErrVersionConflict.Errorf("provided version %s of time interval %s does not match current version %s", mt.Version, mt.Name, curVersion)
+		}
+	} else if mt.Provenance != definitions.Provenance(models.ProvenanceFile) {
+		svc.log.Debug("ignoring optimistic concurrency check because version was not provided", "timeInterval", mt.Name, "operation", "update")
```

Comment:
The logic here is complicated enough to cause me to do a double take when reading, but it's just for a debug level log.

This `else` path is a pretty common case even in API provisioning, so i wonder if we could instead invert the path where we emit the log? i.e.
```
if mt.Version != "" {
    svc.log.Debug("applying optimistic concurrency check because version was provided", ...)
   ...
}
```
which is pretty linear and avoid the extra branch

Reply:
I expect that the optimistic lock is the common path, i.e. normal behavior. Therefore, I decided to emit logs in the negative path to leave some breadcrumbs. It can be useful for investigating why request went through and caused some inconsistent results. Initially, I wanted to emit Warn log but decided to do debug for now because its not supported by any client yet.

Reply:
I refactored code a little bit to make it more readable.